### PR TITLE
[Feat] Add end2end pointpillars & centerpoint(pillar) deployment for mmdet3d

### PR DESCRIPTION
**Pull request by sunjiahao1999**
Thanks for your contribution and we appreciate it a lot. The following instructions would make your pull request more healthy and more easily receiving feedbacks. If you do not understand some items, don't worry, just make the pull request and seek help from maintainers.

## Motivation

Add end2end pointpillars & centerpoint(pillar) deployment for mmdet3d.

## Modification

Rewrite `get_bbox` func for pointpillars and centerpoint.
Rewrite `anchors_single_range`  for pointpillars.
Rewrite `decode` for centerpoint_bbox_coders.

## Checklist

1. Pre-commit or other linting tools are used to fix the potential lint issues.
2. The modification is covered by complete unit tests. If not, please add more unit tests to ensure the correctness.
3. If the modification has a dependency on downstream projects of a newer version, this PR should be tested with all supported versions of downstream projects.
4. The documentation has been modified accordingly, like docstring or example tutorials.

### CODE REVIEW

**Comment by grimoire** [comment]
Did this PR require #1047 ?

**Comment by sunjiahao1999** [comment]
@grimoire Yes, both PRs use the same box3d_nms module

**Comment by RunningLeon** [comment]
@sunjiahao1999 Hi, could you fix conflicts and update this PR? thanks in advance.

**Comment by sunjiahao1999** [comment]
@RunningLeon thanks for reminding me. I will fix it soon.

File: csrc/mmdeploy/backend_ops/tensorrt/batched_bev_nms/trt_batched_bev_nms.cpp [hunk 1]
```diff
@@ -1,6 +1,4 @@
 // Copyright (c) 2021, NVIDIA CORPORATION. All rights reserved.
-// modify from
```
Comment (grimoire) [inline csrc/mmdeploy/backend_ops/tensorrt/batched_bev_nms/trt_batched_bev_nms.cpp:2]: Don't remove the copy right or reference information.

File: mmdeploy/codebase/mmdet3d/deploy/voxel_detection_model.py [hunk 2]
```diff
@@ -93,12 +93,15 @@ def forward(self,
                 'coors': coors
             }
             outputs = self.wrapper(input_dict)
-            result = VoxelDetectionModel.post_process(self.model_cfg,
```
Comment (grimoire) [inline mmdeploy/codebase/mmdet3d/deploy/voxel_detection_model.py:96]: As `post_process` is not used now, should we remove it?

**Comment by grimoire** [comment]
What model did you use to test this PR? Could you provide a script?

File: mmdeploy/codebase/mmdet3d/deploy/voxel_detection_model.py (hunk 2 above)
Comment (sunjiahao1999) [reply mmdeploy/codebase/mmdet3d/deploy/voxel_detection_model.py:96]: ok, I will remove post_process.

File: csrc/mmdeploy/backend_ops/tensorrt/batched_bev_nms/trt_batched_bev_nms.cpp (hunk 1 above)
Comment (sunjiahao1999) [reply csrc/mmdeploy/backend_ops/tensorrt/batched_bev_nms/trt_batched_bev_nms.cpp:2]: ok, I will add it

File: mmdeploy/codebase/mmdet3d/models/anchor3d_head.py [hunk 3]
```diff
@@ -0,0 +1,107 @@
...
+        nms_pre = cfg.get('nms_pre', -1)
+        if nms_pre > 0 and scores.shape[1] > nms_pre:
+            if self.use_sigmoid_cls:
+                max_scores, _ = scores.max(dim=2)
+            else:
+                max_scores, _ = scores[..., :-1].max(dim=2)
+            max_scores = max_scores[0]
+            _, topk_inds = max_scores.topk(nms_pre)
+            anchors = anchors[topk_inds, :]
+            bbox_pred = bbox_pred[:, topk_inds, :]
```
[the remark this answers is not in the document]
Comment (sunjiahao1999) [reply mmdeploy/codebase/mmdet3d/models/anchor3d_head.py:87]: Before `max_scores = max_scores` max_score tensor dimension is `[num_batch, num_classes]`. Here we only consider batches as 1.

**Comment by rkotimi** [comment]
When can this PR be merged? I want to use it.

**Comment by sunjiahao1999** [comment]
@tpoisonooo @grimoire

**Comment by grimoire** [comment]
> What model did you use to test this PR? Could you provide a script?

**Comment by sunjiahao1999** [comment]
test command for pointpillar on trt backend:
```
                "configs/mmdet3d/voxel-detection/voxel-detection_tensorrt_dynamic-kitti-32x4.py",
                "${PATH_TO_MMDET3D}/configs/pointpillars/hv_pointpillars_secfpn_6x8_160e_kitti-3d-3class.py",
                "${PATH_TO_MMDET3D}/checkpoints/hv_pointpillars_secfpn_6x8_160e_kitti-3d-3class_20220301_150306-37dc2420.pth",
                "${PATH_TO_MMDET3D}/demo/data/kitti/000008.bin",
                "--work-dir",
                "mmdeploy-models/pointpillars",
                "--device",
                "cuda:0",
                "--show"
```

**Comment by sunjiahao1999** [comment]
test command for centerpoint(pillar) on trt backend:
```
python tools/deploy.py configs/mmdet3d/voxel-detection/voxel-detection_tensorrt_dynamic-nus-20x5.py \
${PATH_TO_MMDET3D}/configs/centerpoint/centerpoint_02pillar_second_secfpn_4x8_cyclic_20e_nus.py \
${PATH_TO_MMDET3D}/checkpoints/epoch_20.pth \
${PATH_TO_MMDET3D}/data/nuscenes/samples/LIDAR_TOP/n008-2018-08-01-15-16-36-0400__LIDAR_TOP__1533151603547590.pcd.bin \
--work-dir mmdeploy-models/centerpoint \
--device cuda:0 \
--show
```

File: csrc/mmdeploy/backend_ops/tensorrt/batched_bev_nms/trt_batched_bev_nms.hpp [hunk 4]
```diff
@@ -78,4 +78,4 @@ class TRTBatchedBEVNMSCreator : public TRTPluginCreatorBase {
                                             size_t serialLength) TRT_NOEXCEPT override;
 };
 }  // namespace mmdeploy
-#endif  // TRT_BATCHED_NMS_PLUGIN_CUSTOM_H
```
Comment (grimoire) [inline csrc/mmdeploy/backend_ops/tensorrt/batched_bev_nms/trt_batched_bev_nms.hpp:81]: Why do we remove this comment?

Comment (sunjiahao1999) [reply csrc/mmdeploy/backend_ops/tensorrt/batched_bev_nms/trt_batched_bev_nms.hpp:81]: I found `csrc/mmdeploy/backend_ops/tensorrt/batched_rotated_nms/trt_batched_rotated_nms.hpp` also remove this comment. Just follow it.

Comment (sunjiahao1999) [reply csrc/mmdeploy/backend_ops/tensorrt/batched_bev_nms/trt_batched_bev_nms.hpp:81]: Do we need to keep it?

Comment (grimoire) [reply csrc/mmdeploy/backend_ops/tensorrt/batched_bev_nms/trt_batched_bev_nms.hpp:81]: keep it

**Comment by grimoire** [comment]
> 

Failed with data `"${MMDETECTION3D_PATH}/tests/data/nuscenes/samples/LIDAR_TOP/n015-2018-08-02-17-16-37+0800__LIDAR_TOP__1533201470948018.pcd.bin"`

**Comment by grimoire** [comment]
> 

Failed with data `"${MMDETECTION3D_PATH}/tests/data/nuscenes/samples/LIDAR_TOP/n015-2018-08-02-17-16-37+0800__LIDAR_TOP__1533201470948018.pcd.bin"`

**Comment by sunjiahao1999** [comment]
> > 
> 
> Failed with data `"${MMDETECTION3D_PATH}/tests/data/nuscenes/samples/LIDAR_TOP/n015-2018-08-02-17-16-37+0800__LIDAR_TOP__1533201470948018.pcd.bin"`

Sorry to reply to you so late due to some personal matters. The reason for this failure is that the number of points in this samples in mmdet3d/tests is too small (just for some UT). Testing this deployment needs to use more common Nusences samples,

**Review by grimoire — APPROVED** [review]
LGTM

**Pull request merged**